### PR TITLE
Make with-sentry example deployable and runnable

### PR DESCRIPTION
**Pull request by jesstelford**
Ref #1852

I was unable to get the original example working locally, then when I did, I had problems with deploying to production (env vars).

This PR:

- Adds missing dependencies
- Implements better handling of errors thrown in `getInitialProps()`
- Gets it ready for deployment (eg; to `now`) with the right ENV vars setup (`SENTRY_DSN`).
- Keeps setup/config consistent between server/client
- Includes an example custom error page
- Retains consistency between dev & prod
- Doesn't report errors to Sentry from dev (but still dumps them to the console)

with-sentry

cc @amytych @Enalmada @sheerun @Tomekmularczyk @lardissone who have contributed to this example in the past.

### CODE REVIEW

File: examples/with-sentry/next.config.js [hunk 1]
```diff
@@ -1,18 +1,14 @@
...
-      config.plugins.push(
-        new webpack.DefinePlugin({
-          'process.env.SENTRY_DSN': JSON.stringify(SENTRY_DSN),
-          'process.env.SENTRY_RELEASE': JSON.stringify(buildId)
-        })
-      )
-    }
+  webpack: (config, { isServer, buildId }) => {
+    config.plugins.push(
+      new webpack.DefinePlugin({
```
Comment (timneutkens) [inline examples/with-sentry/next.config.js]: This should probably use `env` now that it's part of Next.js.

```
env: {
  SENTRY_DSN: process.env.SENTRY_DSN,
}
```

**Review by timneutkens — APPROVED** [review]

**Comment by karlvd** [comment]
Thanks for providing the update @jesstelford!
I have tried to get it working but without too much success.

To get the example to run express has to be added and some additional changes to package.json.
- Add express as dependency
- Change start to run the custom server + add env variables + set to production

Here is my updated package.json

{
  "name": "with-sentry",
  "version": "1.1.0",
  "scripts": {
    "dev": "cross-env SENTRY_DSN=[From Sentry] SENTRY_TOKEN=[From Sentry] node server.js",
    "build": "next build",
    "start": "cross-env SENTRY_DSN=[From Sentry] SENTRY_TOKEN=[From Sentry] NODE_ENV=production node server.js"
  },
  "dependencies": {
    "@sentry/browser": "^4.6.1",
    "@sentry/node": "^4.6.1",
    "@zeit/next-source-maps": "0.0.4-canary.1",
    "cookie-parser": "1.4.4",
    "js-cookie": "2.2.0",
    "next": "8.0.3",
    "react": "16.8.3",
    "react-dom": "16.8.3",
    "sentry-testkit": "^2.1.0",
    "uuid": "^3.3.2",
    "express": "^4.16.4"
  },
  "devDependencies": {
    "cross-env": "^5.2.0"
  },
  "license": "ISC"
}

When running the example with npm run start (production), the last (server side) test gets reported to Sentry, the first 3 client side tests does not get reported.

I´m on Windows 10 but I doubt it´s an OS issue and env vars get picked up through cross-env. To get the partly working setup I have just cloned the repo with `npx create-next-app --example with-sentry with-sentry-app` and made the adjustments above and started in production mode.

I can see some Sentry related output in the console when the client errors fire but it never gets sent, checked with fiddler as well. Do you have any clues?

**Comment by jesstelford** [comment]
Great catch! I'll open a PR to fix those missing deps / start scripts.

One thing that wasn't noted was the need to set `SENTRY_DSN` during the build phase also:

```
SENTRY_DSN=https://<id>@sentry.io/<id> npm run build
```

Without this, the `SENTRY_DSN` value wont be included in the bundle, no matter what you set on the `env` when running the app.

I'll make this clearer in the example 👍

**Comment by karlvd** [comment]
Thanks for clarifying! Tested and can confirm client-side reporting is working now.
I think there may be another issue with server errors only getting reported once. 

To reproduce I am using the cloned repo with the adjustments.
Test 1-3 gets reported every time to Sentry.
Test 4 gets reported once after server start, subsequent requests are not reaching Sentry. Restarting the server and going the SSR error link trigger a new error in Sentry.

I have tried various checks like adding query strings, incognito mode etc.

**Comment by jesstelford** [comment]
> I think there may be another issue with server errors only getting reported once.

So, this is a _feature_ of the original example: https://github.com/zeit/next.js/pull/5727

If you comment out these lines, you'll see the error come through every time: https://github.com/zeit/next.js/blob/canary/examples/with-sentry/utils/sentry.js#L36-L40

I'm not 100% sure on why deduplication is necessary / how to change the behaviour.

**Comment by sheerun** [comment]
It's not for duplication but grouping of errors, you don't get two groups for each error, one for server, and once for client (Sentry groups errors by stack trace by default).

I don't know if it's possible to group by one thing and deduplicate by another.

**Comment by karlvd** [comment]
I did some more tests, here is the outcome.

**Client side**
Works a expected, every error is sent to Sentry, grouped correctly and incrementing the number for subsequent errors.

**Server side with example conf**
The first error is reported, subsequent errors are not sent to Sentry and does not increment the "Events" number

**Server side with modified conf 1**
Commenting out this row
`//scope.setFingerprint([err.message])`
Same result, first error gets reported and nothing in subsequent errors.

**Server side with modified conf 2**
Forcing a new fingerprint
`scope.setFingerprint([(Math.random() * 10000).toString()])`
Now errors get reported every time but the will of course create a new error in Sentry and grouping is totally broken.

It seems to have to do with deduplication in the client, it can be diabled adding the following to SentryOptions used in Sentry.init

```
 const sentryOptions = {
    integrations: integrations => {
      return integrations.filter(integration => integration.name !== 'Dedupe');
    }
  }

```

The docs says it´s not recommended but it´s a bit confusing that none of the client tests gets deduped but the server test does. Personally I want everything sent to Sentry and have spike protection there instead.

Thanks again for your effort in getting this working!!

**Comment by jesstelford** [comment]
@sheerun Interesting - I think the way the example is structured now, it shouldn't be possible to get the same error on client & server. If the server has an error, it passes an 'error' prop which will short-circuit execution at the highest possible level and instead show the Sentry error screen.

@karlvd Great investigating! I'm of the same mind you are; deduping shouldn't silently drop errors. Instead, I'd prefer to get a flood and selectively ignore them (otherwise, I wouldn't know the relative magnitude of errors!).

I wont have any more time to work on this, but #6521 is merged, so feel free to open a PR to add your investigation results 👍

**Comment by sheerun** [comment]
@jeanbauer I don't mean literally the same error instance but the same error "class", i.e. error that is causes by the same line of code. When you don't set fingerprint you get duplicated *groups* in Sentry, please see @karlvd comment

I think fingerprinting should be left by message on both client and server, but deduplication should be removed from server and left in the client, this way clients won't spam with duplicated errors and we'll have all errors properly grouped in Sentry

**Pull request merged**